### ISSUE TITLE
Compilation fails on Windows

### ISSUE THREAD

**Issue by dboehmer** [robmueller/cache-fastmmap#17]
This is an error report I received from somebody installing my project [Coocook](https://github.com/dboehmer/coocook). Cache::FastMmap is one dependency and fails to build.

Looking at http://matrix.cpantesters.org/?dist=Cache-FastMmap+1.48 I am not sure whether Cache::FastMmap supports Win32. Does it?

How to build and install it?

```
==> Found dependencies: Cache::FastMmap
Searching Cache::FastMmap (1.29) on cpanmetadb ...
--> Working on Cache::FastMmap
Fetching http://www.cpan.org/authors/id/R/RO/ROBM/Cache-FastMmap-1.48.tar.gz
-> OK
Unpacking Cache-FastMmap-1.48.tar.gz
Entering Cache-FastMmap-1.48
Checking configure dependencies from META.json
Checking if you have ExtUtils::MakeMaker 7.00 ... Yes (7.38)
Configuring Cache-FastMmap-1.48
Running Makefile.PL
Checking if your kit is complete...
Looks good
Generating a gmake-style Makefile
Writing Makefile for Cache::FastMmap
Writing MYMETA.yml and MYMETA.json
-> OK
Checking dependencies from MYMETA.json ...
Checking if you have Storable 0 ... Yes (3.15)
Checking if you have ExtUtils::MakeMaker 7.00 ... Yes (7.38)
Building and testing Cache-FastMmap-1.48
cp lib/Cache/FastMmap.pm blib\lib\Cache\FastMmap.pm
Running Mkbootstrap for FastMmap ()
"C:\Strawberry\perl\bin\perl.exe" -MExtUtils::Command -e chmod -- 644 "FastMmap.bs"
"C:\Strawberry\perl\bin\perl.exe" -MExtUtils::Command::MM -e cp_nonempty -- FastMmap.bs blib\arch\auto\Cache\FastMmap\FastMmap.bs 644
"C:\Strawberry\perl\bin\perl.exe" "C:\Strawberry\perl\lib\ExtUtils/xsubpp"  -typemap C:\STRAWB~1\perl\lib\ExtUtils\typemap  FastMmap.xs > FastMmap.xsc
"C:\Strawberry\perl\bin\perl.exe" -MExtUtils::Command -e mv -- FastMmap.xsc FastMmap.c
gcc -c  -I. -s -O2 -DWIN32 -DWIN64 -DCONSERVATIVE -D__USE_MINGW_ANSI_STDIO -DPERL_TEXTMODE_SCRIPTS -DPERL_IMPLICIT_CONTEXT -DPERL_IMPLICIT_SYS -DUSE_PERLIO -fwrapv -fno-strict-aliasing -mms-bitfields -s -O2   -DVERSION=\"1.48\" -DXS_VERSION=\"1.48\"  "-IC:\STRAWB~1\perl\lib\CORE"   FastMmap.c
gcc -c  -I. -s -O2 -DWIN32 -DWIN64 -DCONSERVATIVE -D__USE_MINGW_ANSI_STDIO -DPERL_TEXTMODE_SCRIPTS -DPERL_IMPLICIT_CONTEXT -DPERL_IMPLICIT_SYS -DUSE_PERLIO -fwrapv -fno-strict-aliasing -mms-bitfields -s -O2   -DVERSION=\"1.48\" -DXS_VERSION=\"1.48\"  "-IC:\STRAWB~1\perl\lib\CORE"   mmap_cache.c
gcc -c  -I. -s -O2 -DWIN32 -DWIN64 -DCONSERVATIVE -D__USE_MINGW_ANSI_STDIO -DPERL_TEXTMODE_SCRIPTS -DPERL_IMPLICIT_CONTEXT -DPERL_IMPLICIT_SYS -DUSE_PERLIO -fwrapv -fno-strict-aliasing -mms-bitfields -s -O2   -DVERSION=\"1.48\" -DXS_VERSION=\"1.48\"  "-IC:\STRAWB~1\perl\lib\CORE"   win32.c
win32.c:161:5: error: conflicting types for 'mmc_lock_page'
 int mmc_lock_page(mmap_cache* cache, MU32 p_offset) {
     ^~~~~~~~~~~~~
In file included from win32.c:26:
mmap_cache_internals.h:125:5: note: previous declaration of 'mmc_lock_page' was here
 int mmc_lock_page(mmap_cache* cache, MU64 p_offset);
     ^~~~~~~~~~~~~
win32.c: In function '_mmc_set_error':
win32.c:216:5: warning: 'strncat' specified bound 1024 equals destination size [-Wstringop-overflow=]
     strncat(errbuf, ": ", 1024);
     ^~~~~~~~~~~~~~~~~~~~~~~~~~~
gmake: *** [Makefile:346: win32.o] Error 1
-> FAIL Installing Cache::FastMmap failed. See C:\Users\XXX_USERNAME_XXX\.cpanm\work\1578136759.14308\build.log for details. Retry with --force to force install it.
-> FAIL Installing the dependencies failed: Module 'Cache::FastMmap' is not installed
```

**Comment by dboehmer** [comment]
This is perl 5, version 30, subversion 1 (v5.30.1) built for MSWin32-x64-multi-thread

**Comment by dboehmer** [comment]
I asked on #win32 on irc.perl.org and received this reply from user genio:

> <genio> let me see what happens for me
<genio> same issue here. Type mismatch on the function declaration in the header file vs the implementation in the win32 C file.
<genio> I got it to compile and pass tests by changing the type in the win32.c file on the function. Short of compiling and passing tests, I haven't attempted to see what that possibly messes up
<genio> win32.c on line 161
<genio> int mmc_lock_page(mmap_cache* cache, MU32 p_offset) {    ->   int mmc_lock_page(mmap_cache* cache, MU64 p_offset) {

**Comment by robmueller** [comment]
Unfortunately I don't have easy access perl on windows, but I think I know what the problem is. I've pushed what I think is a correct fix to master. Could you try checking out the latest version from github and seeing if that compiles for you. If so, I'll build and release a new cpan version.

**Comment by h-vn** [comment]
I can confirm that it compiles. The tests also pass apart from test 20 that call umask() and other things that may not work well on Windows. I can look at the test and see if I can make it to do something usual on Windows.

Here'a transcript. First make sure we are up-to-date (git MinGW shell):
```
$ git pull --verbose
From https://github.com/robmueller/cache-fastmmap
 = [up to date]      master     -> origin/master
Already up to date.

$ git status --verbose
On branch master
Your branch is up to date with 'origin/master'.

nothing to commit, working tree clean
```

Then switch to command windows running 64 bit Strawberry Perl 5.30.0, compile and test:
```
C:\work\cache-fastmmap>perl Makefile.PL
Checking if your kit is complete...
Looks good
Generating a gmake-style Makefile
Writing Makefile for Cache::FastMmap
Invalid LICENSE value 'perl' ignored
Writing MYMETA.yml and MYMETA.json

C:\work\cache-fastmmap>gmake
cp lib/Cache/FastMmap.pm blib\lib\Cache\FastMmap.pm
Running Mkbootstrap for FastMmap ()
"C:\sperl64-5.30.0.1\perl\bin\perl.exe" -MExtUtils::Command -e chmod -- 644 "FastMmap.bs"
"C:\sperl64-5.30.0.1\perl\bin\perl.exe" -MExtUtils::Command::MM -e cp_nonempty -- FastMmap.bs blib\arch\auto\Cache\FastMmap\FastMmap.bs 644
"C:\sperl64-5.30.0.1\perl\bin\perl.exe" "C:\sperl64-5.30.0.1\perl\lib\ExtUtils/xsubpp"  -typemap C:\sperl64-5.30.0.1\perl\lib\ExtUtils\typemap  FastMmap.xs > FastMmap.xsc
"C:\sperl64-5.30.0.1\perl\bin\perl.exe" -MExtUtils::Command -e mv -- FastMmap.xsc FastMmap.c
gcc -c  -I. -s -O2 -DWIN32 -DWIN64 -DCONSERVATIVE -D__USE_MINGW_ANSI_STDIO -DPERL_TEXTMODE_SCRIPTS -DPERL_IMPLICIT_CONTEXT -DPERL_IMPLICIT_SYS -DUSE_PERLIO -fwrapv -fno-strict-aliasing -mms-bitfields -s -O2   -DVERSION=\"1.48\" -DXS_VERSION=\"1.48\"  "-IC:\sperl64-5.30.
0.1\perl\lib\CORE"   FastMmap.c
gcc -c  -I. -s -O2 -DWIN32 -DWIN64 -DCONSERVATIVE -D__USE_MINGW_ANSI_STDIO -DPERL_TEXTMODE_SCRIPTS -DPERL_IMPLICIT_CONTEXT -DPERL_IMPLICIT_SYS -DUSE_PERLIO -fwrapv -fno-strict-aliasing -mms-bitfields -s -O2   -DVERSION=\"1.48\" -DXS_VERSION=\"1.48\"  "-IC:\sperl64-5.30.
0.1\perl\lib\CORE"   mmap_cache.c
gcc -c  -I. -s -O2 -DWIN32 -DWIN64 -DCONSERVATIVE -D__USE_MINGW_ANSI_STDIO -DPERL_TEXTMODE_SCRIPTS -DPERL_IMPLICIT_CONTEXT -DPERL_IMPLICIT_SYS -DUSE_PERLIO -fwrapv -fno-strict-aliasing -mms-bitfields -s -O2   -DVERSION=\"1.48\" -DXS_VERSION=\"1.48\"  "-IC:\sperl64-5.30.
0.1\perl\lib\CORE"   win32.c
"C:\sperl64-5.30.0.1\perl\bin\perl.exe" -MExtUtils::Mksymlists \
     -e "Mksymlists('NAME'=>\"Cache::FastMmap\", 'DLBASE' => 'FastMmap', 'DL_FUNCS' => {  }, 'FUNCLIST' => [], 'IMPORTS' => {  }, 'DL_VARS' => []);"
g++.exe FastMmap.def -o blib\arch\auto\Cache\FastMmap\FastMmap.xs.dll -mdll -s -L"C:\sperl64-5.30.0.1\perl\lib\CORE" -L"C:\sperl64-5.30.0.1\c\lib" FastMmap.o mmap_cache.o win32.o   "C:\sperl64-5.30.0.1\perl\lib\CORE\libperl530.a" -lmoldname -lkernel32 -luser32 -lgdi32 -
lwinspool -lcomdlg32 -ladvapi32 -lshell32 -lole32 -loleaut32 -lnetapi32 -luuid -lws2_32 -lmpr -lwinmm -lversion -lodbc32 -lodbccp32 -lcomctl32 -Wl,--enable-auto-image-base
"C:\sperl64-5.30.0.1\perl\bin\perl.exe" -MExtUtils::Command -e chmod -- 755 blib\arch\auto\Cache\FastMmap\FastMmap.xs.dll

C:\work\cache-fastmmap>gmake test
"C:\sperl64-5.30.0.1\perl\bin\perl.exe" -MExtUtils::Command::MM -e cp_nonempty -- FastMmap.bs blib\arch\auto\Cache\FastMmap\FastMmap.bs 644
"C:\sperl64-5.30.0.1\perl\bin\perl.exe" "-MExtUtils::Command::MM" "-MTest::Harness" "-e" "undef *Test::Harness::Switches; test_harness(0, 'blib\lib', 'blib\arch')" t/*.t
t/1.t ... ok
t/10.t .. ok
t/11.t .. ok
t/12.t .. ok
t/13.t .. ok
t/14.t .. ok
t/15.t .. ok
t/16.t .. ok
t/17.t .. ok
t/18.t .. skipped: No FD_CLOEXEC tests, running on Windows
t/19.t .. ok
t/2.t ... ok
t/20.t .. 1/7
#   Failed test 'default persmissions 0640'
#   at t/20.t line 19.
#          got: '438'
#     expected: '416'

#   Failed test 'can set to 0600'
#   at t/20.t line 26.
#          got: '438'
#     expected: '384'
# Looks like you failed 2 tests of 7.
t/20.t .. Dubious, test returned 2 (wstat 512, 0x200)
Failed 2/7 subtests
t/21.t .. 'findmnt' is not recognized as an internal or external command,
operable program or batch file.
t/21.t .. skipped: Large file tests need tmpfs with at least 8G
t/3.t ... ok
t/4.t ... ok
t/5.t ... ok
t/6.t ... skipped: No GTop or /proc/, no memory leak tests
t/7.t ... ok
t/8.t ... ok
t/9.t ... ok

Test Summary Report
-------------------
t/20.t (Wstat: 512 Tests: 7 Failed: 2)
  Failed tests:  3, 5
  Non-zero exit status: 2
Files=21, Tests=529, 34 wallclock secs ( 0.12 usr +  0.03 sys =  0.16 CPU)
Result: FAIL
Failed 1/21 test programs. 2/529 subtests failed.
gmake: *** [Makefile:1053: test_dynamic] Error 255
```

**Comment by robmueller** [comment]
Great. I've just merged your test fix as well (thanks) and released 1.49 to cpan.
